File: Swift/Sorting/sorting algorithms/BubbleSort.swift
//
//  BubbleSort.swift
//  Sort
//
//  Created by Paul Kraft on 21.12.15.
//  Copyright © 2015 Paul Kraft. All rights reserved.
//

extension Sortable {
    /// - complexity: O(n^2)
    public mutating func bubbleSort(by order: (Element, Element) throws -> Bool) rethrows {
        // ...
        try self.bubbleSort(in: self.indices, by: order)
    }
    
    /// - complexity: O(n^2)
    public mutating func bubbleSort(in range: CountableRange<Int>, by order: (Element, Element) throws -> Bool) rethrows {
        // ...
        // an array with 1 element doesn't need to be sorted
        guard range.count > 1 else { return }
        
        // while it is not sorted, didChange will always be true after one cycle
        var didChange = false // reduces best case to O(n), when array is sorted at the beginning, otherwise Theta()
        for j in range.reversed() {
            
            didChange = false
            for i in range.startIndex..<j {
                
                // self[i + 1] < self[i]
                if try order(self[i + 1], self[i]) {
                    didChange = true
                    swap(&self[i], &self[i + 1])
                }
            }
            if !didChange { return }
        }
    }
}

extension Array where Element : Comparable {
    /// complexity: O(n^2)
    public mutating func bubbleSort() {
        self.bubbleSort(by: { $0 < $1 })
    }
}

File: C++/Sorting/sorting algorithms/BubbleSort.cpp
//
//  BubbleSort.cpp
//  Sorting
//
//  Created by Paul Kraft on 08.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#include "BubbleSort.hpp"

BubbleSort::BubbleSort() : Sort("BubbleSort\t\t\t") {}

void BubbleSort::sort(Comparable *array[], int length) {
    sort(array, 0, length - 1);
}

void BubbleSort::sort(Comparable *array[], int left, int right) {
    bool didChange;
    do {
        didChange = false;
        for(int i = left; i < right; i++) {
            if(*array[i+1] < array[i]) {
                didChange = true;
                Comparable *tmp = array[i];
                array[i] = array[i+1];
                array[i+1] = tmp;
            }
        }
    } while(didChange);
}
